### ISSUE TITLE
wallclock timestamp suggests multisecond latencies, monotonic timestamp does not

### ISSUE THREAD

**Issue by cmurf** [systemd/systemd#22554]
**systemd version the issue has been seen with**
 > systemd-250.3-4.fc36.x86_64

**Used distribution**
 > Fedora 36

**Linux kernel version used** (`uname -a`)

 > `5.17.0-0.rc4.96.fc36.x86_64+debug`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > consistent time differential between back to back journal entries

**Unexpected behaviour you saw**
 > wallclock and monotonic time differences differ

wallclock suggests a ~6s time gap between these two messages
```
Feb 17 10:23:29 fovo.local kernel: [drm:drm_mode_addfb2] [FB:130]
Feb 17 10:23:36 fovo.local systemd-journald[610]: Data hash table of /var/log/journal/6f69bf39af5a4dbdbb627b3f3971e626/system.journal has a fill level at 75.0 (174763 of 233016 items, 134217728 >
```
monotonic suggests essentially no time gap between these two messages
```
[   53.999374] fovo.local kernel: [drm:drm_mode_addfb2] [FB:130]
[   54.000891] fovo.local systemd-journald[610]: Data hash table of /var/log/journal/6f69bf39af5a4dbdbb627b3f3971e626/system.journal has a fill level at 75.0 (174763 of 233016 items, 134217728 f>
```

But I'm also seeing many reversals of timestamp, where different times are interleaved, causing some difficulty understanding the actual order of events, in just this one boot there's dozens of such reversals. I don't know if that's related to this problem.

**Steps to reproduce the problem**
> Fedora 36 Workstation, default installation
> BOOT_IMAGE=(hd0,gpt4)/vmlinuz-5.17.0-0.rc4.96.fc36.x86_64+debug root=UUID=b0cdab05-5632-4170-9610-d6ae158f8ec3 ro rootflags=subvol=root3536 mitigations=off enforcing=0 systemd.debug-shell systemd.log_level=debug preempt=full drm.debug=0x06
> journalctl -b vs journalctl -b -o short-monotonic gives a different impression about the gaps/latencies between events.

**Additional program output to the terminal or log subsystem illustrating the issue**

```
2013625-__CURSOR=s=b91b0921b56b472a86d9f93fec33737d;i=93be4;b=1cb023f686dd46d79beece3eba581e6f;m=32f5a3d;t=5d83a0653abe5;x=fdabc411299d6e46
2013626-__REALTIME_TIMESTAMP=1645118609402853
2013627-__MONOTONIC_TIMESTAMP=53434941
2013628-_BOOT_ID=1cb023f686dd46d79beece3eba581e6f
2013629-_SOURCE_MONOTONIC_TIMESTAMP=53999374
2013630-_TRANSPORT=kernel
2013631-PRIORITY=7
2013632-SYSLOG_FACILITY=0
2013633-SYSLOG_IDENTIFIER=kernel
2013634-MESSAGE=[drm:drm_mode_addfb2] [FB:130]
2013635-_MACHINE_ID=6f69bf39af5a4dbdbb627b3f3971e626
2013636-_HOSTNAME=fovo.local
2013637-
2013638-__CURSOR=s=b91b0921b56b472a86d9f93fec33737d;i=93be5;b=1cb023f686dd46d79beece3eba581e6f;m=39c5f88;t=5d83a06c0b130;x=aceb900479288ee8
2013639-__REALTIME_TIMESTAMP=1645118616547632
2013640-__MONOTONIC_TIMESTAMP=60579720
2013641-_BOOT_ID=1cb023f686dd46d79beece3eba581e6f
2013642-_TRANSPORT=kernel
2013643-_MACHINE_ID=6f69bf39af5a4dbdbb627b3f3971e626
2013644-_HOSTNAME=fovo.local
2013645-_SOURCE_MONOTONIC_TIMESTAMP=54000891
2013646-PRIORITY=6
2013647-SYSLOG_FACILITY=5
2013648-SYSLOG_IDENTIFIER=systemd-journald
2013649-SYSLOG_PID=610
2013650:MESSAGE=Data hash table of /var/log/journal/6f69bf39af5a4dbdbb627b3f3971e626/system.journal has a fill level at 75.0 (174763 of 233016 items, 134217728 file size, 767 bytes per hash table item), suggesting rotation.
2013651-_PID=610
2013652-_UID=0
2013653-_GID=0
2013654-_COMM=systemd-journal
2013655-_EXE=/usr/lib/systemd/systemd-journald
2013656-_CMDLINE=/usr/lib/systemd/systemd-journald
2013657-_CAP_EFFECTIVE=25402800cf
2013658-_SELINUX_CONTEXT=system_u:system_r:init_t:s0
2013659-_SYSTEMD_CGROUP=/system.slice/systemd-journald.service
2013660-_SYSTEMD_UNIT=systemd-journald.service
2013661-_SYSTEMD_SLICE=system.slice
2013662-_SYSTEMD_INVOCATION_ID=e2e4cd275051420bb6e737a2675dcc64
2013663-

```

Persisent log is on nvme, KXG6AZNV512G TOSHIBA, io scheduler is none

**Comment by cmurf** [comment]
Three journals in the tar.zst
two original journal files, current and previous; and `journalctl -b -o export`
https://drive.google.com/file/d/16O1U-453Dvr6T0U-4beFl8rx6PEYEpDB/view?usp=sharing

**Comment by cmurf** [comment]
I guess I'm confused about what can/can't be inferred from realtime vs monotonic time. So what's going on is, I'm looking for boot/startup delays, and when I see these multisecond time gaps between events I get suspicious. Like a 6-7s gap here, right as journal rotation is triggered. So I'm just not sure which timestamp I can rely on for that purpose, or even if I can rely on either one.

**Comment by cmurf** [comment]
Also, the above example is (mistakenly) done with an F35 to F36 upgraded system that had nodatacow disabled on /var/log/journal. However, retesting with an actual default clean install of F36, with nodatacow set on active journals, I'm seeing the same issues.

```
Feb 17 12:43:57 fovo.local systemd-machined[834]: varlink: New incoming connection.
Feb 17 12:44:00 fovo.local systemd-journald[608]: Data hash table of /var/log/journal/1ba2f5e7849c488bbba8ed1fa16e0265/system.journal has a fill level at 75.0 (174763 of 233016 items, 134217728 file size, 767 bytes per hash table item), suggesting rotation.
```

```
[   27.990559] fovo.local systemd-machined[834]: varlink: New incoming connection.
[   28.009052] fovo.local systemd-journald[608]: Data hash table of /var/log/journal/1ba2f5e7849c488bbba8ed1fa16e0265/system.journal has a fill level at 75.0 (174763 of 233016 items, 134217728 file size, 767 bytes per hash table item), suggesting rotation.
```

**Comment by DaanDeMeyer** [comment]
So after diving into the logic, what I think is causing the differences between these timestamps is that journalctl is showing timestamps recorded at different times for messages coming from /dev/kmsg depending on whether you use the default output or the monotonic one.

If we look at the output_short() function in logs-show.c, when showing wall clock time or monotonic time, it will always prefer the times from respectively _SOURCE_REALTIME_TIMESTAMP and _SOURCE_MONOTONIC_TIMESTAMP if those fields are present in the message. If those fields are not available in the message, it will fall back to the Entry object's realtime and monotonic time fields respectively.

The difference between the _SOURCE fields and the Entry object time information is that the _SOURCE fields time value is recorded when the message is generated, whereas the Entry object time information is recorded when the message is processed by journald.

Now for log messages coming from /dev/kmsg, _SOURCE_MONOTONIC_TIMESTAMP is included in the kmsg format but _SOURCE_REALTIME_TIMESTAMP isn't, so when showing the default journalctl output, the time displayed is the Entry object's wall clock time (recorded when the message was processed by journald), whereas when using the short-monotonic output, the time displayed is the value from _SOURCE_MONOTONIC_TIMESTAMP (recorded when the kernel generated the log message).

cc @poettering who might also know more.

**Comment by poettering** [comment]
@DaanDeMeyer' analysis looks right to me.

We could try to convert the monotonic timestamp to a realtime one the moment we recieve it. This should usually be quite accurate, except for a short time window when the wallclock is changed, but even in that case it's not going to be totally off...

So maybe we can just fake _SOURCE_REALTIME_TIMESTAMP when we have _SOURCE_MONOTONIC_TIMESTAMP to make this less confusing, even if we don#t actually know the accurate realtime timestamp

**Comment by poettering** [comment]
mm, maybe we should add _SOURCE_REALTIME_TIMESTAMP_CONVERTED as separate field and then make the log viewer use that if _SOURCE_REALTIME_TIMESTAMP is not set.

**Comment by DaanDeMeyer** [comment]
What I'm curious about is why the journald generated wallclock time has such large gaps when the source monotonic time doesn't. This would suggest pretty severe delays in journald processing kernel messages that happened milliseconds from each other.

**Comment by poettering** [comment]
> What I'm curious about is why the journald generated wallclock time has such large gaps when the source monotonic time doesn't. Wouldn't this suggest pretty severe delays in journald processing kmsg log messages that happened milliseconds from each other?

maybe it's when journald flushes stuff from /run to /usr? that usually takes a bit.

**Comment by mwilck** [comment]
> What I'm curious about is why the journald generated wallclock time has such large gaps when the source monotonic time doesn't. Wouldn't this suggest pretty severe delays in journald processing kmsg log messages that happened milliseconds from each other?

In this particular case, my guess is that the 7s difference are caused by journald rotating the journal, and reserving the necessary space in the new journal file. **systemd-journald** is special, because its own log messages always pass through `/dev/kmsg`. So as noted by @DaanDeMeyer, `_SOURCE_MONOTONIC_TIMESTAMP` represents the time at which journald generated, the message, and the other timestamps the time at which it actually wrote it to the journal, which means the message has done a round-trip through the kernel in the meantime.

Looking at the `_SOURCE_MONOTONIC_TIMESTAMP` of PID 610, it seems that it took >6s to reserve the space in the journal, and that indeed journald couldn't process any other log messages during that time.

```
[   54.000891] fovo.local systemd-journald[610]: Data hash table of /var/log/journal/6f69bf39af5a4dbdbb627b3f3971e626/system.journal has a fill level at 75.0 (174763 of>
[   54.000961] fovo.local systemd-journald[610]: /var/log/journal/6f69bf39af5a4dbdbb627b3f3971e626/system.journal: Journal header limits reached or header out-of-date, >
[   54.001040] fovo.local systemd-journald[610]: Rotating...
[   54.019931] fovo.local systemd-journald[610]: Journal effective settings seal=yes keyed_hash=yes compress=yes compress_threshold_bytes=512B
[   54.021624] fovo.local systemd-journald[610]: Creating journal file /var/log/journal/6f69bf39af5a4dbdbb627b3f3971e626/system.journal on a btrfs file system, and copy>
[   54.047564] fovo.local systemd-journald[610]: Reserving 333 entries in field hash table.
[   54.099598] fovo.local systemd-journald[610]: Reserving 233016 entries in data hash table.
[   60.423032] fovo.local systemd-journald[610]: Creating journal file /var/log/journal/6f69bf39af5a4dbdbb627b3f3971e626/user-1000.journal on a btrfs file system, ...
```

**Comment by DaanDeMeyer** [comment]
> Looking at the _SOURCE_MONOTONIC_TIMESTAMP of PID 610, it seems that it took >6s to reserve the space in the journal, and that indeed journald couldn't process any other log messages during that time.

I would think this would be plausible on a system with a HDD or similar hardware, but should it really be taking that long to reserve X MB of space on a modern nvme drive?

**Comment by mwilck** [comment]
> ... should it really be taking that long ...

I don't know if it *should* take so long, but is there another more plausible explanation for the logs of PID 610? If you look at the journal, you see that **journald** actually warns about performance issues caused by btrfs and CoW, so perhaps that's part of the issue. It's also possible that other IO was going on in the system, slowing down journald's operation.

**Comment by mwilck** [comment]
While we are at it: [The linked journal](https://github.com/systemd/systemd/issues/22554#issuecomment-1043321240) demonstrates another issue with `-o short-monotonic`: It isn't monotonic.

```
[ 4274.052063] fovo.local kernel: i915 0000:00:02.0: [drm:__intel_display_power_put_domain [i915]] disabling always-on
[ 4274.089861] fovo.local kernel: ACPI: EC: interrupt blocked
[ 4596.683362] fovo.local kernel: ACPI: EC: interrupt unblocked
[ 4599.484288] fovo.local kernel: i915 0000:00:02.0: [drm:__intel_display_power_get_domain.part.0 [i915]] enabling always-on
...
[ 4600.189351] fovo.local systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/sssd_2dkcm_2esocket interface=org.freedeskt
op.DBus.Properties member=PropertiesChanged cookie=1287 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
[ 4278.347613] fovo.local audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=sssd-kcm comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
```
As the system goes to sleep and wakes up again, the kernel time (dmesg timestamp aka `_SOURCE_MONOTONIC_TIMESTAMP` moves forward, while `CLOCK_MONOTONIC` does not. From that point on, kernel time stamps are ~320s ahead of `__MONOTONIC_TIMESTAMP`, causing the impression that "monotonic" time jumps back and forth.

@poettering, IMHO this indicates that `journalctl -o short-monotonic` wrongly makes the  assumption that the dmesg timestamp represents the same time as `CLOCK_MONOTONIC`. That's generally not the case (`printk()` timestamps use `local_clock()`, which is different from `ktime_get()`). While the specific issue here, caused by suspend, [might be fixed](https://linux.die.net/man/2/clock_gettime) by using `CLOCK_BOOTIME`, I have seen other cases where dmesg time stamps and `CLOCK_MONOTONIC` diverged without suspend.

**Comment by YHNdnzj** [comment]
This is discussed again in https://github.com/systemd/systemd/pull/33386 and other issues/PRs linked to it. Indeed kmsgs are in CLOCK_BOOTTIME.
